Building master on gentoo: QA Warnings

### Zig Version

0.12.0-dev.1840+d787b78d2

### Steps to Reproduce and Observed Behavior

```
 * QA Notice: Package triggers severe warnings which indicate that it
 *            may exhibit random runtime failures.
 * /var/no-tmpfs/portage/dev-lang/zig-9999/work/zig-9999/build/zig2.c:3378343: warning: assignment to 'const uint16_t (*)[16]' {aka 'const short unsigned int (*)[16]'} from incompatible pointer type 'const uint16_t *' {aka 'const short unsigned int *'} [-Wincompatible-pointer-types]
 * /var/no-tmpfs/portage/dev-lang/zig-9999/work/zig-9999/build/zig2.c:1493587: warning: 'memcpy' reading 16 bytes from a region of size 10 [-Wstringop-overread]
 * /var/no-tmpfs/portage/dev-lang/zig-9999/work/zig-9999/build/zig2.c:2744871: warning: 'memcpy' reading 16 bytes from a region of size 10 [-Wstringop-overread]
 * /var/no-tmpfs/portage/dev-lang/zig-9999/work/zig-9999/build/zig2.c:2744945: warning: 'memcpy' reading 16 bytes from a region of size 10 [-Wstringop-overread]

 * Please do not file a Gentoo bug and instead report the above QA
 * issues directly to the upstream developers of this software.
 * Homepage: https://ziglang.org/
```

My apologizes if totally meaningless.

 

### Expected Behavior

No QA warnings reported

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -24311,7 +24311,8 @@ fn analyzeMinMax(
 
 fn upgradeToArrayPtr(sema: *Sema, block: *Block, ptr: Air.Inst.Ref, len: u64) !Air.Inst.Ref {
     const mod = sema.mod;
-    const info = sema.typeOf(ptr).ptrInfo(mod);
+    const ptr_ty = sema.typeOf(ptr);
+    const info = ptr_ty.ptrInfo(mod);
     if (info.flags.size == .One) {
         // Already an array pointer.
         return ptr;
@@ -24330,10 +24331,11 @@ fn upgradeToArrayPtr(sema: *Sema, block: *Block, ptr: Air.Inst.Ref, len: u64) !A
             .address_space = info.flags.address_space,
         },
     });
-    if (info.flags.size == .Slice) {
-        return block.addTyOp(.slice_ptr, new_ty, ptr);
-    }
-    return block.addBitCast(new_ty, ptr);
+    const non_slice_ptr = if (info.flags.size == .Slice)
+        try block.addTyOp(.slice_ptr, ptr_ty.slicePtrFieldType(mod), ptr)
+    else
+        ptr;
+    return block.addBitCast(new_ty, non_slice_ptr);
 }
 
 fn zirMemcpy(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!void {

--- a/src/link/C.zig
+++ b/src/link/C.zig
@@ -136,6 +136,8 @@ pub fn deinit(self: *C) void {
     self.string_bytes.deinit(gpa);
     self.fwd_decl_buf.deinit(gpa);
     self.code_buf.deinit(gpa);
+    self.lazy_fwd_decl_buf.deinit(gpa);
+    self.lazy_code_buf.deinit(gpa);
 }
 
 pub fn freeDecl(self: *C, decl_index: InternPool.DeclIndex) void {

--- a/test/behavior/memcpy.zig
+++ b/test/behavior/memcpy.zig
@@ -66,6 +66,27 @@ fn testMemcpyDestManyPtr() !void {
     try expect(buf[4] == 'o');
 }
 
+test "@memcpy slice" {
+    if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest;
+    if (builtin.zig_backend == .stage2_arm) return error.SkipZigTest;
+    if (builtin.zig_backend == .stage2_sparc64) return error.SkipZigTest;
+
+    try testMemcpySlice();
+    try comptime testMemcpySlice();
+}
+
+fn testMemcpySlice() !void {
+    var buf: [5]u8 = undefined;
+    const dst: []u8 = &buf;
+    const src: []const u8 = "hello";
+    @memcpy(dst, src);
+    try expect(buf[0] == 'h');
+    try expect(buf[1] == 'e');
+    try expect(buf[2] == 'l');
+    try expect(buf[3] == 'l');
+    try expect(buf[4] == 'o');
+}
+
 comptime {
     const S = struct {
         buffer: [8]u8 = undefined,

